commercetools-convenience/src/main/java/io/sphere/sdk/queries/QueryAllImpl.java

```diff
@@ -22,42 +24,140 @@ private QueryAllImpl(final QueryDsl<T, C> baseQuery, final long pageSize) {
+                         .mapToObj(page -> queryPage(client, page)
+                             .thenApply(PagedQueryResult::getResults)
+                             .thenApply(callback)
+                             .toCompletableFuture())
+                         .collect(toList());
+    }
+
+    /**
+     * Given a {@link Consumer}, this method applies this consumer on each page of the results from
+     * {@link this} instance's {@code baseQuery}.
+     *
+     * @param client   the CTP client that the query is run on.
+     * @param consumer the consumer that gets called on each page of results.
+     * @return an empty future.
+     */
+    @Nonnull
+    CompletionStage<Void> run(final SphereClient client, final Consumer<List<T>> consumer) {
+        return queryPage(client, 0).thenCompose(result -> {
+            final List<CompletableFuture<Void>> futureResults = new ArrayList<>();
+            consumer.accept(result.getResults());
+            futureResults.addAll(queryNextPages(client, result.getTotal(), consumer));
+            return CompletableFuture.allOf(futureResults.toArray(new CompletableFuture[futureResults.size()]));
+        });
+    }
+
+    /**
+     * Given a {@link Consumer}, this method first calculates the total numbr of pages resulting from the query, then
+     * it applies this consumer on each page of the results from {@link this} instance's {@code baseQuery}
+     * and returns a list of futures after applying of this consumer on each page.
+     *
+     * @param client        the CTP client that the query is run on.
+     * @param totalElements the total number of elements resulting from the query.
+     * @param consumer      the consumer to apply on each page of results.
+     * @return a list of futures of applying the consumer on each page of results.
+     */
+    @Nonnull
+    private List<CompletableFuture<Void>> queryNextPages(final SphereClient client, final long totalElements,
+                                                         final Consumer<List<T>> consumer) {
+        final long totalPages = getTotalNumberOfPages(totalElements);
+        return LongStream.range(1, totalPages)
+                         .mapToObj(page -> queryPage(client, page)
+                             .thenApply(PagedQueryResult::getResults)
+                             .thenAccept(consumer)
+                             .toCompletableFuture())
+                         .collect(toList());
+    }
+
+    /**
+     * Given a total number of elements {@code totalElements} this method calculates the number of pages needed that
+     * would be needed to cover all the elements with set {@code pageSize} of {@link this} instance.
+     *
+     * @param totalElements number of elements to get the number of pages for.
+     * @return the total number of pages.
+     */
+    long getTotalNumberOfPages(final long totalElements) {
+        return (long) Math.ceil((double) totalElements / pageSize);
```

Comment:
In Java the `/` operator has a higher predence that the cast operator. That means that this code performs an integer division and afterwards does the cast.
The correct code looks like: `Math.ceil(((double) totalElements) / pageSize)`

Would be nice if you could add some more test cases for this method 😉

Reply:
> In Java the / operator has a higher predence that the cast operator. 

actually, no, [cast has higher priority](http://introcs.cs.princeton.edu/java/11precedence/)

Reply:
[But don't re-invent the wheel](https://www.google.de/search?client=opera&q=java+pagination+calculate+number+pages&oq=java+pagination+calculate+number+pages&gs_l=psy-ab.3...274434.274841.0.275577.5.5.0.0.0.0.244.376.0j1j1.2.0....0...1.1.64.psy-ab..4.0.0._N9H518yYzw):

```java
int pages = (totalElements + pageSize - 1) / pageSize;
```

Reply:
@andrii-kovalenko-ct Not sure how is it better? You will have to round in both implementations.

Reply:
the question not about rounding: this is just less confusing, has less casting and no `Math` method calls.

But talking precisely: in the last proposition there is no rounding, it is completely integer arithmetics.